### ISSUE TITLE
[3.5.4->4.0 Regression] "no signature found for builtin" when using autodoc on pybind11-based code

### ISSUE THREAD

**Issue by inducer** [sphinx-doc/sphinx#9200]
**Describe the bug**

Sphinx 3.5.4 completes these steps without a problem ([build log](https://gitlab.tiker.net/inducer/islpy/-/jobs/265216))

```sh
python -m venv islpy-doc-test
source islpy-doc-test/bin/activate
git clone --recursive https://github.com/inducer/islpy.git
cd islpy
git checkout 0aa174e6956778df0c9e674f83cfb5e0e1d206b2
python setup.py install  # ignore the pybind11 warning, build takes about two minutes
cd doc
pip install sphinx sphinx_copybutton furo
pip install --upgrade sphinx  # furo advertises it needs sphinx < 4, but it works just fine with 4.0.0
make html SPHINXOPTS="-W --keep-going -n"
```

Sphinx 4.0.0 crashes with [this traceback](https://gist.github.com/inducer/fbd5b7202ac9158cf91f5d3cd177780f). ([build log](https://github.com/inducer/islpy/pull/59/checks?check_run_id=2540516783)).

Here's a snippet from the build log, since Github will eventually delete it:
```
Exception occurred:
  File "/opt/hostedtoolcache/Python/3.9.5/x64/lib/python3.9/inspect.py", line 2145, in _signature_from_builtin
    raise ValueError("no signature found for builtin {!r}".format(func))
ValueError: no signature found for builtin <instancemethod name at 0x7f4d02df0280>
The full traceback has been saved in /tmp/sphinx-err-dctonk0f.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

**Additional context**

#9199

**Comment by MatthijsBurgh** [comment]
This bug is introduced in #8983. Is there a reason why in `autodoc/__init__.py` a TypeError is catched, but a ValueError is reraised? As in other parts a ValueError is not reraised.

As it is a change in the `PropertyDocumenter` it only occurs in properties.

**Comment by vallsv** [comment]
Hi, here is an exception with Python 3.7.

It looks like it come from the same initial problem.
```
Exception occurred:
  File "/opt/conda/envs/docenv/lib/python3.7/inspect.py", line 2396, in _signature_from_callable
    raise ValueError('callable {!r} is not supported by signature'.format(obj))
ValueError: callable operator.itemgetter(4) is not supported by signature
```

**Comment by exhuma** [comment]
We ran into the same issue today. For those looking for a workaround, you can simply skip the `4.0.0` version by adding `!= 4.0.0` in your dependencies.

For example:

```
# setup.py
...
setup(
    ...
    install_requires=[
        "sphinx >= 1.7, != 3.0.2, != 3.2.0, != 4.0.0",
        ...
```

```
# pyproject.toml w/ poetry
...
[tool.poetry.dev-dependencies]
sphinx = ">= 1.7, != 3.0.2, != 3.2.0, != 4.0.0"
...
```

By simply blacklisting broken versions, you will still get any future upgrades (as opposed to pinning it as `<4.0`).

**Comment by MatthijsBurgh** [comment]
Fixed by #9190. So can be closed IMO

**Comment by tk0miya** [comment]
Thank you for reporting. Same as https://github.com/sphinx-doc/sphinx/issues/9189. Closing.
